race-buffer/src/lib.rs

```diff
@@ -1,162 +1,234 @@
+    type Output = Self;
 
-    impl Entry for OrderedEntry {
-        fn is_prefix(&self) -> bool {
-            assert!(!self.is_suffix());
-            self.is_prefix_unchecked()
-        }
+    fn sub(self, rhs: Self) -> Self::Output {
+        (Into::<u64>::into(self) - Into::<u64>::into(rhs)).into()
     }
+}
 
-    impl OrderedEntry {
-        const PREFIX_TAG: u32 = 0x80000000; // 2^31
-        const SUFFIX_TAG: u32 = 0x40000000; // 2^30
-
-        pub(crate) fn to_index(&self) -> u32 {
-            if self.is_suffix() {
-                self.0 - Self::SUFFIX_TAG
-            } else if self.is_prefix_unchecked() {
-                self.0 - Self::PREFIX_TAG
-            } else {
-                self.0
-            }
-        }
+impl Add<u64> for SeqNum {
+    type Output = Self;
 
-        pub(crate) fn from_index(i: u32) -> Self {
-            assert!(i <= Self::SUFFIX_TAG);
-            OrderedEntry(i)
-        }
+    fn add(self, rhs: u64) -> Self::Output {
+        (Into::<u64>::into(self) + rhs).into()
+    }
+}
 
-        pub(crate) fn from_index_prefix(i: u32) -> Self {
-            assert!(i <= Self::SUFFIX_TAG);
-            OrderedEntry(i + Self::PREFIX_TAG)
-        }
+impl AddAssign<u64> for SeqNum {
+    fn add_assign(&mut self, rhs: u64) {
+        *self = *self + rhs
+    }
+}
 
-        pub(crate) fn from_index_suffix(i: u32) -> Self {
-            assert!(i <= Self::SUFFIX_TAG);
-            OrderedEntry(i + Self::SUFFIX_TAG)
-        }
+impl Sub<u64> for SeqNum {
+    type Output = Self;
 
-        pub(crate) fn is_prefix_unchecked(&self) -> bool {
-            self.0 >= Self::PREFIX_TAG
-        }
+    fn sub(self, rhs: u64) -> Self::Output {
+        (Into::<u64>::into(self) - rhs).into()
+    }
+}
 
-        pub(crate) fn is_suffix(&self) -> bool {
-            self.0 >= Self::SUFFIX_TAG && self.0 < Self::PREFIX_TAG
-        }
+/// Returns the corresponding index in backing storage to given sequence number
+#[inline]
+fn get_seqn_index(storage_cap: usize, seqn: SeqNum, use_base_2_indexing: bool) -> usize {
+    // Cast to usize safe because index in storage slice cannot be greater than usize
+    if use_base_2_indexing {
+        // Index is lowest n bits of seqn where storage_cap is 2^n
+        (Into::<u64>::into(seqn) & (storage_cap as u64 - 1)) as usize
+    } else {
+        (Into::<u64>::into(seqn) % storage_cap as u64) as usize
+    }
+}
 
-        // Invariant: Check if entries all have correct index
-        fn entries_correct_index(rbuf: &[Option<OrderedEntry>]) -> bool {
-            for (idx, entry) in rbuf.iter().enumerate().filter(|(_, e)| e.is_some()) {
-                if entry.unwrap().to_index() != idx as u32 {
-                    return false;
-                }
-            }
-            return true;
-        }
+/// Calculate the number of entries missed based on the read and overwrite sequence numbers
+#[inline]
+fn num_missed(read_seqn: SeqNum, overwrite_seqn: SeqNum) -> SeqNum {
```

Comment:
I don't think this should return a `SeqNum`, which I understand is largely present for indexing and comparison between a few specific cursors, which is not how we would want someone to use the outcome of this function.  A `u64` seems appropriate.

Reply:
Nevermind, I found the places where it is being used for cursor incrementing, and that seems fine enough.